### ISSUE TITLE
Remove 'document_type' setting

### ISSUE THREAD

**Issue by jordansissel** [logstash-plugins/logstash-output-elasticsearch#259]
We've been advising users (for years!) to use `document_type` for different schemas of logs on the same index. However, recently, I have learned that this is not a correct way to use Elasticsearch!

TODO: Add link/references to this from the Elasticsearch repo

Proposal: We should remove the `document_type` setting, and update our docs to encourage using different index names for different (and possibly conflicting) document schemas within Elasticsearch.

**Comment by AlexRRR** [comment]
@jordansissel I am just about to implement `document_type` care to share why this is a bad idea? could not find anything on this?

**Comment by untergeek** [comment]
@AlexRRR, Elasticsearch still requires a `_type` for each document, and Logstash provides this via the `type` field automatically, or via the `document_type` setting.

What @jordansissel is referring to are the changes since Elasticsearch 2.0 which disallow conflicting field mappings within an index, even if they exist within different `_type`s.  Earlier versions of Elasticsearch allowed for this, and this was sub-optimal.

Please see [mapping changes](https://www.elastic.co/guide/en/elasticsearch/reference/2.1/breaking_20_mapping_changes.html#_conflicting_field_mappings).  The blog post you linked has a lighter explanation of the same.

**Comment by tuananhtd** [comment]
Sorry guys, in the path of finding how to use the `document_type` I landed here. I'm just started playing with ELK stack a few days. And my newbie's thought is that if I have different type of event to log I should use different `type`.
I could didn't using the `document_type` setting yet and now I just find out we have `type` and `_type` what is the difference then?
Hope to hear your thought @untergeek

**Comment by untergeek** [comment]
@tuananhtd 

As previously mentioned, Elasticsearch requires a `_type`.  Logstash uses the value of the `type` field to provide Elasticsearch with a value for `_type`.  If there is no `type` in a Logstash event, Logstash will automatically provide Elasticsearch with a value of `logs` as the value for `_type`.

The `document_type` setting simply overrides whatever Logstash would have set and provides the value to Elasticsearch.

As far as the difference is concerned, Elasticsearch uses `_type` as part of document mapping.  Every document can be directly discovered if you know the index name, the `_type`, and the document id.  Using a `type` field in Elasticsearch, you can more easily distinguish between different data sets.  However, the most current recommendation for Elasticsearch is to use a single data type (as in a single mapping, or `_type`) per index for the reasons mentioned in https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/259#issuecomment-178064836 

This can make it seem like having a `type` field redundant. Regardless, Logstash frequently needs this `type` in order to route events through specific filters, and/or to the correct destinations.  At present, you could tag/type your fields using the `@metadata` field in Logstash, as nothing in that object will propagate to Elasticsearch.

**Comment by untergeek** [comment]
@tuananhtd This is why @jordansissel is recommending we remove `document_type` altogether, and perhaps even use `@metadata` fields for event routing.  It will clean things up in Logstash _and_ in Elasticsearch as well.

**Comment by tuananhtd** [comment]
@untergeek  thank you so much for the response!

As a result of my debugging I find out that I just cannot use the field `type` in my JSON object. However change the field's name will make it work. After read your comments I think may be the [mapping changes](https://www.elastic.co/guide/en/elasticsearch/reference/2.1/breaking_20_mapping_changes.html#_conflicting_field_mappings) that you mentioned is the reason. Besides that the `type` field (without the underscore) is always set to `log`. `document_type` setting will only affect the `_type` field.

Regarding to **single data type per index** I'm just curious that will it generate to much overhead since each index will have multiple shards so based on my naive though each query will have a overhead as   `(number of shards) * (number of types)`?

**Comment by andrewvc** [comment]
I still think users should get this level of control. We *should* discourage users from using `document_type`, but there's no reason to not allow them to use it IMHO.

**Comment by andrewvc** [comment]
I'll mention its also necessary for parent/child docs.

**Comment by arn-k** [comment]
Omitting document_type in a config causes documents to be created in Elasticsearch with type "doc". Yet the Elasticsearch documentation for the planned removal of type recommends using "_doc" for type to aid the transition to the removal of types.

Should this plugin default the type to "_doc" instead of "doc" when document_type is omitted?

**Comment by Dris101** [comment]
I think that is correct. I am doing a re-indexing from 5.6 to 6.4 and if I want logstash to use the now recommended "_doc" rather than "doc" (see https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html) I am having to use `document_type => "_doc"`, which is itself deprecated

**Comment by arn-k** [comment]
Yes, same for us, which means we get deprecation warnings in our logs about using document_type. Would gladly remove it IF the default was _doc instead of doc.
